src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexPrefixAnalyzer.cs

```diff
@@ -12,6 +11,329 @@ namespace System.Text.RegularExpressions
     /// <summary>Detects various forms of prefixes in the regular expression that can help FindFirstChars optimize its search.</summary>
     internal static class RegexPrefixAnalyzer
     {
+        /// <summary>Cache of ToString() strings for the ASCII chars.</summary>
+        /// <remarks>The strings are lazily created on first use.</remarks>
+        private static string[]? s_asciiCharStrings;
+
+        /// <summary>Gets the ToString() string for the specified char.</summary>
+        private static string GetCharString(char ch)
+        {
+            // If the character isn't ASCII, just ToString it.
+            if (ch >= 128)
+            {
+                return ch.ToString();
+            }
+
+            // Use a lazily-initialized cache of strings for ASCII characters. The overall cache is initialized
+            // with Interlocked.CompareExchange simply to avoid accidentally throwing out a lot of strings accidentally.
+
+            string[] asciiCharString =
+                s_asciiCharStrings ??
+                Interlocked.CompareExchange(ref s_asciiCharStrings, new string[128], null) ??
+                s_asciiCharStrings;
+
+            return asciiCharString[ch] ??= ch.ToString();
+        }
+
+        /// <summary>Finds an array of multiple prefixes that a node can begin with.</summary>
+        /// <param name="node">The node to search.</param>
+        /// <returns>
+        /// If a fixed set of prefixes is found, such that a match for this node is guaranteed to begin
+        /// with one of those prefixes, an array of those prefixes is returned.  Otherwise, null.
+        /// </returns>
+        public static string[]? FindPrefixes(RegexNode node)
+        {
+            // Minimum string length for prefixes to be useful. If any prefix has length 1,
+            // then we're generally better off just using IndexOfAny with chars.
+            const int MinPrefixLength = 2;
+
+            // Arbitrary string length limit to avoid creating strings that are longer than is useful and consuming too much memory.
+            // Some of the algorithms used also create strings by appending a single character at a time, which is O(N^2) in the
+            // final length of the string, so keeping this small helps avoid problematic complexity.
+            const int MaxPrefixLength = 8;
+
+            // Limit based on what IndexOfAny is able to handle most efficiently. This can be updated in future if/when IndexOfAny improves further.
```

Comment:
Should this point to the private constant in IndexOfAny that defines the value so that it aids in keeping the two values in sync?

Reply:
Which constant are you referring to?

Reply:
Sorry, so long since I reviewed this and now I'm trying to page back what I was thinking and can't recall. Should be fine to ignore, I believe I was just trying to suggest to add a pointer here to where that limit is set.